### PR TITLE
TST: Added test for groupby dropping NA when grouping on columns

### LINKED ISSUE #21755: Some groupby methods keep null grouping keys
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({'A': ['a','b',None,'a','b',None], 'B': range(6)})
df.groupby('A').sum()  # output has two rows
df.groupby('A').first() # output has two rows
df.groupby('A').head(1) # output has three rows, one for the null group
df.groupby('A').nth(1) # output has three rows, but one of them has the wrong value in the 'A' column.
```
#### Problem description

My understanding is that when grouping on a column, pandas excludes null values ([link](http://pandas.pydata.org/pandas-docs/stable/groupby.html#na-and-nat-group-handling)). However, when calling the `head` method on a GroupBy object, the null group is returned. While I generally would support keeping null values across the board, this inconsistency contradicts the documentation.  Looking to see if similar methods had the same behavior, I discovered that `nth` has an even worse bug.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-45-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: 3.4.2
pip: 10.0.1
setuptools: 38.4.0
Cython: None
numpy: 1.14.3
scipy: 1.1.0
pyarrow: 0.9.0
xarray: None
IPython: 6.2.1
sphinx: 1.7.3
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: 0.4.0
matplotlib: 2.1.2
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by pbhoopala**
- [x] closes #21755
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -2840,3 +2840,15 @@ def test_groupby_index_name_in_index_content(val_in, index, val_out):
...
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_on_column_drops_na():
+    # GH 21755
+    df = DataFrame(
+        {"a": [np.nan, "foo", "bar", np.nan, np.nan, "baz"], "b": [0, 1, 2, 3, 4, 5]}
+    )
+    result = df.groupby("a").head(1)
+    expected = DataFrame({"a": ["foo", "bar", "baz"], "b": [1, 2, 5]})
```
Comment (phofl) [inline pandas/tests/groupby/test_groupby.py:2851]: Can you define the index in here as well? No need to set in an extra line.

Also please merge main

Comment (pbhoopala) [reply pandas/tests/groupby/test_groupby.py]: I just added a commit which defines the index in the DataFrame constructor, as requested. Also, I have merged the pull request to main.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! Can you also test nth? It should have the same output, so can just parameterize the test, e.g.

    @pytest.mark.parametrize("method", ["head", "nth"])

**Comment by rhshadrach** [comment]
Closing as stale. @pbhoopala - if you're interested in continuing just comment here and happy to reopen.

**Pull request closed without merging**